### PR TITLE
Feature: add skills field

### LINKED ISSUE #893: Enhance Create Task Form(according to new requirements)
# Enhance Create Task Form

## Issue Description
Enhance the create task form by removing specific fields and adding new ones as per the updated requirements.

## Current Behavior
The current create task form includes the following fields:
image

## Expected Behavior
The create task form should be updated as follows:

### Fields to be added:
- Skills (multi-select, similar to label selection in Jira / assigning labels to PR on GitHub)
- Assigned to (optional, search user + autocomplete)

### Fields to be removed:
- Task Level
- Category
- Level
- Feature URL
- Feature/Group

All other existing fields should be retained.

After following changes it should look like this ( Not functional right now )

create-task-form2

### when ` status ` is ` available `
 - Do not show ` Assigned to ` feild 
 - Do not show ` Ends on ` feild

## Reproducibility
- [x] This issue is reproducible
- [ ] This issue is not reproducible

## Steps to Reproduce
1. Navigate to the create task form
2. Observe the current fields present in the form
3. Compare with the new requirements

## Severity/Priority
- [ ] Critical
- [ ] High
- [x] Medium
- [ ] Low

## Additional Information
The new 'Skills' field should allow multi-select functionality. 
The 'Assigned to' field should right now only be search with autocomplete functionality.

## Steps to complete this task 
 - [ ]  Add `Skills` feild 
 - [ ]  Remove unecessary feilds
 - [ ] Modify `Assiged to ` field to match the requirements
 
## Checklist
- [x] I have read and followed the project's code of conduct.
- [x] I have searched for similar issues before creating this one.
- [x] I have provided all the necessary information to understand and reproduce the issue.
- [x] I am willing to contribute to the resolution of this issue.

### PR DESCRIPTION
**Pull request by sunilk429**
## Developer Name: Sunil Kumar

---

## Issue Ticket Number

- Closes #893 
## Description
Added Skills multi-select component

## Depends On 
 - Linked PR #914 

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [x] Yes
- [ ] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

</details>

add-skills
image

https://github.com/user-attachments/assets/800b99ac-431f-4232-9aee-ec3a1f5d3670

## Test Coverage
<details>
<summary>Screenshot 1</summary>

all-tests
image

</details>

### CODE REVIEW

File: task/index.html [hunk 1]
```diff
@@ -97,6 +97,43 @@
...
+                class="dropdown-icon"
+              >
+                <path d="m6 9 6 6 6-6"></path>
+              </svg>
+            </button>
+            <div class="popover">
+              <input type="text" class="search-input" placeholder="Search..." />
+              <div class="options-list">
+                <div class="option" data-value="select-all">
+                  <span class="option-label">(Select aAll)</span>
```
Comment (AnujChhikara) [inline task/index.html:129]: Should it be '(Select All)' instead of '(Select aAll)'?

File: task/script.js [hunk 2]
```diff
@@ -1,9 +1,292 @@
...
+  async initializeSkills() {
+    try {
+      const skills = await this.fetchSkills();
+      this.options = skills.map((skill) => ({
+        value: skill.id.toString(),
+        label: skill.name,
+      }));
+      this.init();
+    } catch (error) {
+      alert(`Error Initializing skills: ${error}`);
```
Comment (AnujChhikara) [inline task/script.js:61]: why are we not using toast ?

File: __tests__/tasks/task-dev-dependency.test.js [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    beforeEach(async () => {
+      await page.goto('http://localhost:8000/task?dev=true');
+      await page.waitForNetworkIdle();
+    });
+
+    it('should hide feature URL field', async () => {
+      const featureUrlField = await page.$eval(
+        '#featureUrl',
+        (el) => window.getComputedStyle(el.closest('.inputBox')).display,
+      );
```
[the remark this answers is not in the document]
Comment (sunilk429) [reply __tests__/tasks/task-dev-dependency.test.js:98]: These fields and tests has to be removed later, i can use data-testId for this though, isnt it fine? since this has anyways later to be removed. let me know if i still should use data-testId which by the way i have used for skills field which is new addition

File: __tests__/tasks/task-dev-dependency.test.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+      expect(featureUrlField).toBe('none');
+    });
+
+    it('should hide feature/group radio buttons', async () => {
+      const featureRadio = await page.$eval(
+        '.feature',
+        (el) => window.getComputedStyle(el.closest('.inputBox')).display,
+      );
```
[the remark this answers is not in the document]
Comment (sunilk429) [reply __tests__/tasks/task-dev-dependency.test.js:106]: okay

File: task/index.html (hunk 1 above)
Comment (sunilk429) [reply task/index.html:129]: typo, will fix it

File: task/script.js (hunk 2 above)
Comment (sunilk429) [reply task/script.js:61]: There was no toast been used in case of failure or submission, I mostly took inspiration from the already written code where alert has been used for calling other api's

File: __tests__/tasks/task-dev-dependency.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunilk429) [reply __tests__/tasks/task-dev-dependency.test.js:98]: Because those fields has to be removed... right now they are under feature flag (you can read more in issue description), creating a seperate file is okay

[the remark this answers is not in the document]
Comment (AnujChhikara) [reply __tests__/tasks/task-dev-dependency.test.js:98]: who told you to create separate test file for these changes? Why these tests can't be written in task-dependency.test.js file?

[the remark this answers is not in the document]
Comment (sunilk429) [reply __tests__/tasks/task-dev-dependency.test.js:98]: No one told me, What's the issue though? It can be written in same file as well

[the remark this answers is not in the document]
Comment (AnujChhikara) [reply __tests__/tasks/task-dev-dependency.test.js:98]: if we can write tests in the already present file than we don't need to create a separate file for it.

[the remark this answers is not in the document]
Comment (sunilk429) [reply __tests__/tasks/task-dev-dependency.test.js:98]: I see the concern i have reviewed your PR, if your PR get merged first i will resolve it since you are exporting base_url from mock_data/constants or vice versa. about separate file concern if you still insisit on making changes in same file... let me know i will do it

File: task/script.js [hunk 5]
```diff
@@ -1,9 +1,285 @@
...
+
+if (isDev && taskLevelDiv) {
+  taskLevelDiv.style.display = 'none';
+}
+const skillComponentDiv = document
+  .querySelector('.multi-select-button')
+  ?.closest('.inputBox');
+if (!isDev && skillComponentDiv) {
+  skillComponentDiv.style.display = 'none';
+}
```
Comment (yesyash) [inline task/script.js:34]: - instead of defining a variable and writing a condition what if we define all the variables first and then add the required checks? this will make it easy for others contributing to the codebase to skim the code. what do you think?

File: task/script.js [hunk 6]
```diff
@@ -1,9 +1,285 @@
...
+  featureUrlField.style.display = 'none';
+}
+const featureRadio = document.getElementById('feature')?.closest('.inputBox');
+
+if (isDev && featureRadio) {
+  featureRadio.style.display = 'none';
+}
+const taskLevelDiv = document
+  .querySelector('.inputBox label[for="taskLevel"]')
+  .closest('.inputBox');
```
Comment (yesyash) [inline task/script.js:24]: - instead of using a nested query to select a div can we target the div directly? this will help us prevent any unnecessary changes if something inside the inputBox changes and help us get rid of the `closest` method. what do you think?

File: task/script.js [hunk 7]
```diff
@@ -1,9 +1,285 @@
...
+    try {
+      const response = await fetch(SKILLS_API, {
+        credentials: 'include',
+      });
+      return await response.json();
+    } catch (error) {
+      alert(`Error fetching skills: ${error}`);
+      return [];
+    }
+  }
```
Comment (yesyash) [inline task/script.js:76]: - let's move this outside of the class as it is not a property of the class, also this will allow us to  reuse this method without instantiating the `MultiSelect` class.

File: task/style.css [hunk 8]
```diff
@@ -279,3 +276,148 @@ footer {
...
+/* Multi-select container styles */
+.multi-select-container {
+  position: relative;
+  width: 100%;
+}
+
+.multi-select-button {
+  display: flex;
+  align-items: center;
+  min-height: 48px;
```
Comment (yesyash) [inline task/style.css:289]: - as the height of the button will not be changed let's use `height` instead of min height.

File: task/style.css [hunk 9]
```diff
@@ -279,3 +276,148 @@ footer {
...
+  max-height: 320px;
+  padding: 4px;
+  overflow-y: auto;
+}
+
+.option {
+  color: #101828;
+  display: flex;
+  align-items: center;
+  justify-content: space-between; /* This moves checkbox to right */
```
Comment (yesyash) [inline task/style.css:369]: - remove this comment

File: task/script.js (hunk 5 above)
Comment (sunilk429) [reply task/script.js]: > * instead of defining a variable and writing a condition what if we define all the variables first and then add the required checks? this will make it easy for others contributing to the codebase to skim the code. what do you think?

You are right, have updated accordingly.

File: task/script.js (hunk 6 above)
Comment (sunilk429) [reply task/script.js]: Have done this, please do review again

File: task/script.js (hunk 7 above)
Comment (sunilk429) [reply task/script.js]: Done

File: task/style.css (hunk 8 above)
Comment (sunilk429) [reply task/style.css:289]: This is different from the regular button... we show "Skill badges" on button which can increase the height depending on number of badges

File: task/style.css (hunk 9 above)
Comment (sunilk429) [reply task/style.css]: Done

**Review by yesyash — DISMISSED** [review]

**Review by yesyash — DISMISSED** [review]

**Review by iamitprakash — APPROVED** [review]

**Review by Achintya-Chatterjee — COMMENTED** [review]
tests are failing , kindly look into this .

**Comment by sunilk429** [comment]
> tests are failing , kindly look into this .

This PR had conflicts, fixed incorrectly... Will fix it once I am back. Currently I am ooo

File: __tests__/tasks/task-dependency.test.js [hunk 10]
```diff
@@ -1,10 +1,13 @@
+const SKILL_TREE_BACKEND_BASE_URL =
+  'https://services.realdevsquad.com/skilltree/v1';
```
Comment (iamitprakash) [inline __tests__/tasks/task-dependency.test.js:2]: why this url is hardcoded here

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: __tests__/tasks/task-dependency.test.js (hunk 10 above)
Comment (sunilk429) [reply __tests__/tasks/task-dependency.test.js]: > why this url is hardcoded here

Fixed it, please review this.

**Review by iamitprakash — APPROVED** [review]
